homotopy-core/src/diagram.rs

```diff
@@ -220,6 +220,38 @@ impl Diagram {
             Self::DiagramN(d) => d.suspend(s, t),
         }
     }
+
+    pub fn atomic_generator(&self) -> Option<Generator> {
+        match self {
+            Self::Diagram0(d) => Some(d.generator),
+            Self::DiagramN(d) if d.cospans().is_empty() => d.source().atomic_generator(),
+            Self::DiagramN(d) if d.cospans().len() == 1 => d
+                .source()
+                .rewrite_forward(&d.cospans()[0].forward)
+                .unwrap()
+                .atomic_generator(),
+            Self::DiagramN(_) => None,
+        }
+    }
+
+    pub fn atomic_boundaries(&self) -> Option<(Generator, Generator)> {
+        if let Self::DiagramN(d) = self {
+            match (d.source().atomic_generator(), d.target().atomic_generator()) {
```

Comment:
You might want to use [zip](https://doc.rust-lang.org/std/option/enum.Option.html#method.zip).